#include "client/gui/mainwindow.hpp"
#include "ui_mainwindow.h"
#include <QMessageBox>
#include <QInputDialog>
#include <QMenuBar>
#include <QMenu>
#include <QDebug>
#include <QSettings>
#include <QListWidgetItem> 
#include "client/gui/userlistwindow.hpp"
#include "client/gui/createRoom.hpp"
#include "client/chat_client.hpp"
#include "client/gui/groupchatwindow.hpp"
#include <memory>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), ui(new Ui::MainWindow) {
    ui->setupUi(this);
    ui->chatDisplay->setReadOnly(true);
    setupMenuBar();
    setupConnections();
    connectToServer();
}

MainWindow::~MainWindow() {
    delete ui;
    if (chatClient) {
        chatClient->stop();
    }
}

void MainWindow::setupMenuBar() {
    QMenuBar* menuBar = new QMenuBar(this);
    setMenuBar(menuBar);

    QMenu* roomMenu = menuBar->addMenu(tr("ì±„íŒ…ë°©"));
    QAction* roomListAction = roomMenu->addAction(tr("ê·¸ë£¹ì±„íŒ… ì—´ê¸°"));
    connect(roomListAction, &QAction::triggered, this, &MainWindow::showGroupChat);

    QMenu* userMenu = menuBar->addMenu(tr("ìœ ì €"));
    QAction* userListAction = userMenu->addAction(tr("ìœ ì € ëª©ë¡"));
    connect(userListAction, &QAction::triggered, this, &MainWindow::requestUserList);
}

void MainWindow::setupConnections() {
    connect(ui->sendButton, &QPushButton::clicked, this, &MainWindow::sendMessage);
    connect(ui->messageInput, &QLineEdit::returnPressed, this, &MainWindow::sendMessage);
    connect(ui->userListButton, &QPushButton::clicked, this, &MainWindow::requestUserList);
    connect(ui->createRoomButton, &QPushButton::clicked, this, &MainWindow::createNewRoom);
    connect(ui->groupChatButton, &QPushButton::clicked, this, &MainWindow::showGroupChat);
    connect(ui->userListButton2, &QPushButton::clicked, this, &MainWindow::requestUserList);
    connect(ui->mainChatButton, &QPushButton::clicked, this, &MainWindow::showMainChat);
    connect(ui->roomListWidget, &QListWidget::itemClicked, this, &MainWindow::joinSelectedRoom);
}

void MainWindow::connectToServer() {
    QSettings settings;
    QString savedNickname = settings.value("nickname").toString();
    bool ok;

    QString nickname;
    if (savedNickname.isEmpty()) {
        nickname = QInputDialog::getText(this, "ë‹‰ë„¤ì„ ì…ë ¥", "ì‚¬ìš©í•  ë‹‰ë„¤ì„ì„ ì…ë ¥í•˜ì„¸ìš”:", QLineEdit::Normal, "", &ok);
        if (!ok || nickname.isEmpty()) {
            QMessageBox::critical(this, "ì˜¤ë¥˜", "ë‹‰ë„¤ì„ì„ ì…ë ¥í•´ì•¼ í•©ë‹ˆë‹¤.");
            close();
            return;
        }
        // ë‹‰ë„¤ì„ ì €ì¥
        settings.setValue("nickname", nickname);
    } else {
        nickname = QInputDialog::getText(this, "ë‹‰ë„¤ì„ ì…ë ¥", "ì‚¬ìš©í•  ë‹‰ë„¤ì„ì„ ì…ë ¥í•˜ì„¸ìš”:", QLineEdit::Normal, savedNickname, &ok);
        if (!ok) {
            close();
            return;
        }
        if (nickname != savedNickname) {
            // ë‹‰ë„¤ì„ì´ ë³€ê²½ëœ ê²½ìš° ì €ì¥
            settings.setValue("nickname", nickname);
        }
    }

    QString serverIp = QInputDialog::getText(this, "ì„œë²„ IP ì…ë ¥", "ì„œë²„ IP ì£¼ì†Œë¥¼ ì…ë ¥í•˜ì„¸ìš”:", QLineEdit::Normal, "localhost", &ok);
    if (!ok) {
        QMessageBox::critical(this, "ì˜¤ë¥˜", "ì„œë²„ IPë¥¼ ì…ë ¥í•´ì•¼ í•©ë‹ˆë‹¤.");
        close();
        return;
    }

    chatClient = std::make_unique<ChatClient>();
    if (!chatClient->connect(serverIp.toStdString(), 8080)) {
        QMessageBox::critical(this, "ì—°ê²° ì˜¤ë¥˜", "ì„œë²„ì— ì—°ê²°í•  ìˆ˜ ì—†ìŠµë‹ˆë‹¤.");
        close();
        return;
    }

    chatClient->setMessageCallback([this](const std::string& message) {
        appendMessage(QString::fromStdString(message));
    });

    chatClient->start();
    ui->statusLabel->setText("ì—°ê²°ë¨");
    chatClient->sendMessage("/nickname " + nickname.toStdString());
}

void MainWindow::sendMessage() {
    QString message = ui->messageInput->text().trimmed();
    if (!message.isEmpty()) {
        chatClient->sendMessage(message.toStdString());
        ui->messageInput->clear();
    }
}

void MainWindow::appendMessage(const QString& message) {
    QMetaObject::invokeMethod(this, [this, message]() {
        if (message.startsWith("USER_LIST:")) {
            QStringList users = message.mid(10).split(",", Qt::SkipEmptyParts);
            if (!userListWindow) {
                userListWindow = std::make_unique<UserListWindow>(this);
            }
            userListWindow->updateUserList(users);
            userListWindow->show();
            userListWindow->raise();
            userListWindow->activateWindow();
        } else if (message.startsWith("ROOM_LIST:")) {
            QStringList rooms = message.mid(10).split(",", Qt::SkipEmptyParts);
            ui->roomListWidget->clear();
            ui->roomListWidget->addItems(rooms);
        } else if (message.startsWith("ROOM_MSG:")) {
            QString roomMsg = message.mid(9);
            if (groupChatWindow) {
                groupChatWindow->appendMessage(roomMsg);
            }
        } else {
            ui->chatDisplay->append(message);
        }
    }, Qt::QueuedConnection);
}

void MainWindow::requestUserList() {
    if (chatClient) chatClient->sendMessage("/users");
}

// ğŸ”½ ê·¸ë£¹ì±„íŒ… ìŠ¬ë¡¯ êµ¬í˜„
void MainWindow::showGroupChat() {
    if (chatClient) {
        chatClient->sendMessage("/list_rooms");
    }
    ui->stackedWidget->setCurrentWidget(ui->groupChatWidget);  // âœ… ê·¸ë£¹ì±„íŒ… ìœ„ì ¯ìœ¼ë¡œ ì „í™˜
}
void MainWindow::showMainChat() {
    ui->stackedWidget->setCurrentWidget(ui->mainChatWidget);
}

void MainWindow::createNewRoom() {
    CreateRoomDialog dialog(this);
    if (dialog.exec() == QDialog::Accepted) {
        QString roomName = dialog.getRoomName().trimmed();
        QString password = dialog.getPassword().trimmed();
        bool isPrivate = dialog.isPrivate();

        if (roomName.isEmpty()) {
            QMessageBox::warning(this, "ê²½ê³ ", "ë°© ì´ë¦„ì„ ì…ë ¥í•˜ì„¸ìš”.");
            return;
        }

        QString command = QString("/create_room %1").arg(roomName);
        if (isPrivate && !password.isEmpty()) {
            command += QString(" --private --password %1").arg(password);
        }

        currentRoomName = roomName;
        chatClient->sendMessage(command.toStdString());

        // âœ… ìƒˆë¡œ ìƒì„±í•œ ë°© ëª©ë¡ ìš”ì²­
        chatClient->sendMessage("/list_rooms");

        // âœ… ìƒˆ ì°½ ì—´ê¸°
        if (!groupChatWindow) {
            groupChatWindow = std::make_unique<GroupChatWindow>(this);
            connect(groupChatWindow.get(), &GroupChatWindow::sendMessageRequested, this, &MainWindow::sendGroupMessage);
        }

        groupChatWindow->setRoomTitle(roomName);
        groupChatWindow->show();
    }
}

void MainWindow::joinSelectedRoom(QListWidgetItem* item) {
    if (!item) return;
    QString roomName = item->text();
    currentRoomName = roomName;

    chatClient->sendMessage(QString("/join_room %1").arg(roomName).toStdString());

    if (!groupChatWindow) {
        groupChatWindow = std::make_unique<GroupChatWindow>(this);
        connect(groupChatWindow.get(), &GroupChatWindow::sendMessageRequested, this, &MainWindow::sendGroupMessage);
    }

    groupChatWindow->setRoomTitle(roomName);
    groupChatWindow->show();
}

void MainWindow::sendGroupMessage(const QString& msg) {
    if (msg.isEmpty() || currentRoomName.isEmpty()) return;

    QString command = QString("/room_msg %1 %2").arg(currentRoomName).arg(msg);
    chatClient->sendMessage(command.toStdString());
}